### PR TITLE
feat(service-proxy): add service mixin

### PR DESCRIPTION
**Pull request by bajtos**
Implement "ServiceMixin" for applications. This mixin enhances component
registration so that service providers exported by a component are
automatically registered for dependency injection; and adds a new sugar
API for registering service providers manually:

```ts
app.serviceProvider(MyServiceProvicer);
```

The method name "serviceProvider" was chosen deliberately to make it
clear that we are binding a Provider, not a class constructor. Compare
this to `app.repository(MyRepo)` that accepts a class construct. In the
future, we may add `app.service(MyService)` method if there is enough
user demand.

This pull request is the first step towards automated service registration via boot, see #1439.

## Checklist

- [x] `npm test` passes on your machine
- [x] New tests added or existing tests modified to cover all changes
- [x] Code conforms with the [style guide](http://loopback.io/doc/en/contrib/style-guide.html)
- [x] API Documentation in code was updated
- [x] Documentation in [/docs/site](../tree/master/docs/site) was updated
- [x] Affected artifact templates in `packages/cli` were updated
- [x] Affected example projects in `examples/*` were updated

### CODE REVIEW

File: docs/site/Calling-other-APIs-and-Web-Services.md [hunk 1]
```diff
@@ -60,19 +60,6 @@ Install the REST connector used by the new datasource:
...
-```ts
-import {Context} from '@loopback/context';
-
-const context = new Context();
-context.bind('dataSources.geoService').to(ds);
-```
-
-**NOTE**: Once we start to support declarative datasources with
-`@loopback/boot`, the datasource configuration files can be dropped into
-`src/datasources` to be discovered and bound automatically.
```
Comment (bajtos) [inline docs/site/Calling-other-APIs-and-Web-Services.md]: This doc fix is not strictly related to the changes made by my pull request, the section should have been removed when the datasource booter was added. Since it's just a small change, I didn't feel like creating a new commit or a new pull request.

File: packages/service-proxy/src/mixins/service.mixin.ts [hunk 2]
```diff
@@ -0,0 +1,198 @@
...
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Provider} from '@loopback/context';
+import {Application} from '@loopback/core';
+
+/**
+ * Interface for classes with `new` operator.
+ */
+export interface Class<T> {
```
Comment (raymondfeng) [inline packages/service-proxy/src/mixins/service.mixin.ts]: Should we somehow leverage the `Class` type from https://github.com/strongloop/loopback-next/blob/master/packages/context/src/value-promise.ts#L13 or refactor it into a common package?

File: packages/service-proxy/src/mixins/service.mixin.ts [hunk 3]
```diff
@@ -0,0 +1,198 @@
...
+    }
+
+    /**
+     * Get an instance of a component and mount all it's
+     * services. This function is intended to be used internally
+     * by component()
+     *
+     * @param component The component to mount services of
+     */
+    mountComponentRepository(component: Class<{}>) {
```
Comment (raymondfeng) [inline packages/service-proxy/src/mixins/service.mixin.ts]: Shouldn't it be named as `mountComponentServiceProviders`?

File: packages/service-proxy/src/mixins/service.mixin.ts [hunk 4]
```diff
@@ -0,0 +1,198 @@
...
+}
+
+/**
+ * Interface for an Application mixed in with ServiceMixin
+ */
+export interface AppWithServices extends Application {
+  // tslint:disable-next-line:no-any
+  serviceProvider<S>(provider: Class<Provider<S>>): void;
+  component(component: Class<{}>): void;
+  mountComponentRepository(component: Class<{}>): void;
```
Comment (raymondfeng) [inline packages/service-proxy/src/mixins/service.mixin.ts]: It should be `mountComponentServiceProviders` too.

File: packages/service-proxy/src/mixins/service.mixin.ts [hunk 5]
```diff
@@ -0,0 +1,198 @@
...
+        }
+      }
+    }
+  };
+}
+
+/**
+ * Interface for an Application mixed in with ServiceMixin
+ */
+export interface AppWithServices extends Application {
```
Comment (raymondfeng) [inline packages/service-proxy/src/mixins/service.mixin.ts]: `ApplicationWithServices`?

File: packages/service-proxy/src/mixins/service.mixin.ts [hunk 6]
```diff
@@ -0,0 +1,198 @@
...
+  public component(component: Class<{}>) {}
+
+  /**
+   * Get an instance of a component and mount all it's
+   * services. This function is intended to be used internally
+   * by component()
+   *
+   * @param component The component to mount services of
+   */
+  mountComponentRepository(component: Class<{}>) {}
```
Comment (raymondfeng) [inline packages/service-proxy/src/mixins/service.mixin.ts]: mountComponentServiceProviders

File: packages/service-proxy/test/unit/mixin/service.mixin.unit.ts [hunk 7]
```diff
@@ -0,0 +1,97 @@
...
+// This file is licensed under the MIT License.
+// License text available at https://opensource.org/licenses/MIT
+
+import {Application, Component, Provider} from '@loopback/core';
+import {expect} from '@loopback/testlab';
+import {Class, ServiceMixin} from '../../../';
+
+// tslint:disable:no-any
+
+describe('RepositoryMixin', () => {
```
Comment (raymondfeng) [inline packages/service-proxy/test/unit/mixin/service.mixin.unit.ts]: It should be `ServiceProviderMixin`.

File: packages/service-proxy/src/mixins/service.mixin.ts (hunk 2 above)
Comment (bajtos) [reply packages/service-proxy/src/mixins/service.mixin.ts]: Having a shared definition of `Class` and `Constructor` types would be definitely nice.

However, I am not sure if such small typedefs are enough to justify the cost of creating a new package?

Also please note that in `RepositoryMixin` (which I based my work on), `Class` is coming from a different package - see https://github.com/strongloop/loopback-next/blob/f68a45ced9a392286d484b72cc690a6599aeb0c0/packages/repository/src/common-types.ts#L11-L19

Can we leave this improvement out of scope of this pull request please?

File: packages/service-proxy/src/mixins/service.mixin.ts (hunk 3 above)
Comment (bajtos) [reply packages/service-proxy/src/mixins/service.mixin.ts]: Good point, I'll apply both suggestions.

File: packages/service-proxy/src/mixins/service.mixin.ts (hunk 4 above)
Comment (bajtos) [reply packages/service-proxy/src/mixins/service.mixin.ts]: Actually, I prefer to call this method `mountComponentServices`. In the future, if we add support for `app.service(serviceCtor: Class)`, then I'd like `mountComponentServices` to handle both service classes and providers.

File: packages/service-proxy/src/mixins/service.mixin.ts (hunk 2 above)
Comment (raymondfeng) [reply packages/service-proxy/src/mixins/service.mixin.ts]: Sure. +1 to have a separate PR.

File: packages/service-proxy/test/unit/mixin/service.mixin.unit.ts (hunk 7 above)
Comment (bajtos) [reply packages/service-proxy/test/unit/mixin/service.mixin.unit.ts]: See https://github.com/strongloop/loopback-next/pull/1649#discussion_r213006676, I prefer to use `ServiceMixin` instead of `ServiceProviderMixin`. 

In my mind, the mixin is adding APIs for working with services. The fact that services are registered via Providers is just an implementation detail to me.

**Comment by bajtos** [comment]
@raymondfeng comments addressed in 767c54c, LGTY now?

**Comment by marioestradarosa** [comment]
Nice work @bajtos . From the app developer perspective in the current way of registering the services and the proposed way of registering the provider and then injecting the service in the controller I can see a better design. However, can you highlight if there is a performance gain in the proposed way?. And finally what about provider discovery? , to avoid the registration of the provider itself ?

**Comment by bajtos** [comment]
@marioestradarosa thank you for chiming in.

> Can you highlight if there is a gain in performance in this way?

I am not aware of any performance gains. My goal here is to make it easier for LB4 app developers to register their services (service providers). 

> And finally I believe the next step would be to have the provider discovery , to avoid the registration of the provider, imagine an app with a lot of providers ?

Exactly! My next step is to implement a Booter in `@loopback/boot` package that will scan project files, discover service providers and register them via `app.serviceProvider`.

See #1439 for a bigger picture.

**Comment by bajtos** [comment]
@virkt25 thank you for pointing out opportunities to make the code cleaner! I addressed your first two comments in 6b7fa83
